File: batteryMPC.jl
using JuMP
using GLPK

function mpc_run(NL, battInit)
    # NL represents the 24h net load forecast
    # T is the horizon (96 in case of 15 min data)
    # battInit is the battery SoC at the time when this function is called

    T = length(NL) # size(NL)[1]
    #model = Model(solver=GLPKSolverMIP())
    model = Model(with_optimizer(GLPK.Optimizer))

    # Declare variables
    @variable(model, u[t = 1:T], Bin)
    @variable(model, v[t = 1:T], Bin)
    @variable(model, 0 <= Pch[t = 1:T] <= 5 ) # Charging power of battery
    @variable(model, 0 <= Pdis[t = 1:T] <= 5 ) # Disharging power of battery
    @variable(model, 0 <= PfrGrid[t = 1:T] <= 5 ) # Power from grid
    @variable(model, 0 <= PtoGrid[t = 1:T] <= 5 ) # Power to grid
    @variable(model, 0.5 <= Energy[t = 1:T] <= 7.5) # start is a guess, not fixed

    # Declare constraints
    @constraint(model, powerBalance[t = 1:T],
                    NL[t] == PfrGrid[t] - PtoGrid[t] - Pch[t] + Pdis[t])
    @constraint(model, batteryEnergy[t = 2:T],
                    Energy[t] == Energy[t-1] + 0.25*Pch[t-1] - 0.25*Pdis[t-1])
    @constraint(model, batteryInitial[t = 1], Energy[1] == battInit)
    @constraint(model, binVars[t = 1:T], u[t] + v[t] == 1)      # Binary variables for constraints below
    @constraint(model, binCH[t = 1:T], Pch[t] <= u[t] * 5)      # Avoid arbitrage and discharging
    @constraint(model, binDIS[t = 1:T], Pdis[t] <= v[t] * 5)    # battery energy to the grid.
    @constraint(model, bin2GR[t = 1:T], PtoGrid[t] <= u[t] * 5)

    @objective(model, Min, sum(PfrGrid + PtoGrid))
    #solve(model)
    optimize!(model)
    return round(JuMP.value(Pch[1]),digits=2),
           round(JuMP.value(Pdis[1]),digits=2),
           round(JuMP.value(PfrGrid[1]),digits=2),
           round(JuMP.value(PtoGrid[1]),digits=2)
           #round(JuMP.value(Energy[1]),digits=2),
           #round(getobjectivevalue(model),digits=2)
end

File: batteryMPC.py
'''
Author: Dennis van der Meer
E-mail: denniswillemvandermeer[at]gmail.com

This script contains the smart charging algorithm based on model predictive
control. 
'''

import cvxpy as cp
import numpy as np

def mpc(NL,SoC):
    '''
    - NL is the net load forecast and should be numpy array
    - SoC is the initial battery state of charge, i.e., the initial state
    - The function outputs charge and discharge power as well as power to
      and from the grid.
    '''
    T = len(NL)
    n = 1 # States
    m = 4 # Controls
    d = 1 # Disturbances (net load in this case)
    # The following gains are only used in the dynamic equation
    A = 1 # State gain (scalar because n=1)
    B = np.array([[0.25, -0.25, 0, 0]]) # Control gain with delta t
    C = 0 # Disturbance gain (scalar because d=1)
    x_0 = 7.2*SoC # Initial battery capacity

    # Set up the optimization problem:
    x = cp.Variable((n, T+1)) # State
    u = cp.Variable((m, T)) # Controls
    r = cp.Variable((2, T), boolean=True) # Binary variable
    d = cp.Parameter(shape=(NL.shape), value=NL) # Disturbance
    '''
    Herein, control vector u is the following:
    u[0,t] = Pch
    u[1,t] = Pdis
    u[2,t] = PfrGrid
    u[3,t] = PtoGrid
    '''
    # Add constraints and objective function "cost"
    cost = 0
    constr = []
    for t in range(T):
        cost += cp.sum(u[2:4,t])
        constr += [x[:,t+1] == A*x[:,t] + B@u[:,t] + C*d[t],
                   d[t] == u[1,t] - u[0,t] + u[2,t] - u[3,t],
                   u[:,t] >= 0.0,
                   u[0,t] <= 5*r[0,t],
                   u[1,t] <= 5*r[1,t],
                   r[0,t] + r[1,t] <= 1,
                   u[3,t] <= 5*r[0,t],
                   x[:,t] >= 1.44, x[:,t] <= 7.0]
    constr += [x[:,0] == x_0]
    # Problem definition:
    problem = cp.Problem(cp.Minimize(cost), constr)
    problem.solve(solver=cp.GLPK_MI)
    # Return results:
    return(np.transpose(np.around(u.value[:,0], decimals=2)))
